### PR TITLE
Dot versus Colon 2

### PR DESCRIPTION
**Pull request by Perryvw**
Counter proposal to #109 and #108

### CODE REVIEW

**Comment by Perryvw** [comment]
I actually remembered why we use `:` for static methods: when using `this` in a static method it usually refers to the class itself. If we look at what happens in lua this is perfectly preserved by using the `:` syntax and adding a 'self' parameter.

These other ones indeed seem like issues and I will have a look at how to resolve those when I have time.

**Comment by apemanzilla** [comment]
Hmm, that does make sense for `static`, although there are again cases where APIs expect to be called using dot syntax, which complicates things.

**Comment by Perryvw** [comment]
@apemanzilla @lolleko tried to solve by transpiling lambda properties with `.` and methods with `:`. Have a look

File: src/Transpiler.ts [hunk 1]
```diff
@@ -917,8 +917,12 @@ export class LuaTranspiler {
                 return this.transpileArrayCallExpression(node);
             }
 
-            if (expType.symbol && (expType.symbol.flags & ts.SymbolFlags.Namespace)) {
+            const expType2 = this.checker.getTypeAtLocation(node.expression);
```
Comment (apemanzilla) [inline src/Transpiler.ts]: It might be worth renaming `expType` and `expType2` to something more descriptive - such as `containerType` and `functionType` respectively.

Comment (lolleko) [reply src/Transpiler.ts]: agreed

File: test/translation/ts/dotColonFunctionCalls.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+declare const classInstance: TestClass;
+declare const interfaceInstance: TestInterface;
+
+classInstance.colonMethod();
+classInstance.dotMethod();
+interfaceInstance.colonMethod();
+interfaceInstance.dotMethod();
+TestNameSpace.dotMethod();
+TestNameSpace.dotMethod2();
```
Comment (lolleko) [inline test/translation/ts/dotColonFunctionCalls.ts]: Not sure how much sense it makes but maybe add some functional tests too. Probably not that important.

**Comment by apemanzilla** [comment]
Is there anything still holding this back?

**Comment by apemanzilla** [comment]
Found another minor edge case. If you have an instance of an interface extending from two interfaces defining the same method as an arrow function and regular function, it will use the first parent interface to determine the function type. I can't think of any cases where this really matters, though.

```typescript
declare interface A2 {
    method: () => void;
}

declare interface B2 {
    method(): void;
}

declare interface Thing extends B2, A2 {}

declare const thing: Thing;

thing.method();
```

**Comment by Perryvw** [comment]
@apemanzilla I will merge this once I added some functional tests and had a quick look if there is any easy way to deal with union types.

**Comment by apemanzilla** [comment]
Thank you!
